data/provider/src/main/java/tech/pegasys/artemis/api/ValidatorDataProvider.java

```diff
@@ -53,6 +80,96 @@ public ValidatorDataProvider(ValidatorCoordinator validatorCoordinator) {
+      final ValidatorDutiesRequest validatorDutiesRequest) {
+
+    if (validatorDutiesRequest == null || !combinedChainDataClient.isStoreAvailable()) {
+      return SafeFuture.failedFuture(new ChainDataUnavailableException());
+    }
+    final Optional<Bytes32> optionalBlockRoot = combinedChainDataClient.getBestBlockRoot();
+    if (optionalBlockRoot.isEmpty()) {
+      return completedFuture(List.of());
+    }
+
+    UnsignedLong epoch = validatorDutiesRequest.epoch;
+    UnsignedLong slot = BeaconStateUtil.compute_start_slot_at_epoch(epoch);
+    final Bytes32 headBlockRoot = optionalBlockRoot.get();
+    return combinedChainDataClient
+        .getStateAtSlot(slot, headBlockRoot)
+        .thenApply(
+            state -> getValidatorDutiesFromState(state.get(), validatorDutiesRequest.pubkeys))
+        .exceptionally(err -> List.of());
+  }
+
+  @VisibleForTesting
+  protected List<ValidatorDuties> getValidatorDutiesFromState(
+      final tech.pegasys.artemis.datastructures.state.BeaconState state,
+      final List<BLSPubKey> pubKeys) {
+    final List<ValidatorDuties> dutiesList = new ArrayList<>();
+
+    final List<CommitteeAssignment> committees =
+        combinedChainDataClient.getCommitteesFromState(state, state.getSlot());
+    final UnsignedLong firstSlot = state.getSlot();
+    Map<BLSPubKey, List<UnsignedLong>> proposers = new HashMap<>();
+
+    for (int i = 0; i < Constants.SLOTS_PER_EPOCH; i++) {
+      final UnsignedLong thisSlot = firstSlot.plus(UnsignedLong.valueOf(i));
+      BLSPublicKey publicKey = validatorCoordinator.getProposerForSlot(state, thisSlot);
+      BLSPubKey pubkey = new BLSPubKey(publicKey.toBytes());
+      if (proposers.containsKey(pubkey)) {
+        List<UnsignedLong> proposalSlots = proposers.get(pubkey);
+        proposalSlots.add(thisSlot);
+      } else {
+        proposers.put(pubkey, List.of(thisSlot).stream().collect(Collectors.toList()));
+      }
+    }
+
+    for (final BLSPubKey pubKey : pubKeys) {
+      final Integer validatorIndex = getValidatorIndex(state.getValidators().asList(), pubKey);
+      if (validatorIndex == null) {
+        dutiesList.add(new ValidatorDuties(pubKey, null, null, List.of()));
+      } else {
+        List<UnsignedLong> proposedSlots = proposers.getOrDefault(pubKey, List.of());
+        dutiesList.add(
+            new ValidatorDuties(
+                pubKey,
+                validatorIndex,
+                getCommitteeIndex(committees, validatorIndex),
+                proposedSlots));
+      }
+    }
+    return dutiesList;
+  }
+
+  @VisibleForTesting
+  protected static Integer getValidatorIndex(
+      final List<tech.pegasys.artemis.datastructures.state.Validator> validators,
+      final BLSPubKey publicKey) {
+    Optional<tech.pegasys.artemis.datastructures.state.Validator> optionalValidator =
+        validators.stream()
+            .filter(
+                v -> Bytes48.fromHexString(publicKey.toHexString()).equals(v.getPubkey().toBytes()))
```

Comment:
can probably use a bytes constructor here, haven't looked at bytes48 and publickey specifically, but most of those structures are allowing it